share/p2p/shrexnd/client.go

```diff
@@ -0,0 +1,173 @@
+		shares, err := c.getSharesByNamespace(
+			ctx, root, nID, peer)
+		if err != nil {
+			log.Debugw("client: peer returned err", "peer_id", peer.String(), "err", err)
+			continue
+		}
+
+		return shares, nil
+	}
+
+	return nil, errNotAvailable
+}
+
+// getSharesByNamespace gets shares with Merkle tree inclusion proofs from remote host
+func (c *Client) getSharesByNamespace(
+	ctx context.Context,
+	root *share.Root, nID namespace.ID,
+	peerID peer.ID,
+) (share.NamespacedShares, error) {
+	stream, err := c.host.NewStream(ctx, peerID, ndProtocolID)
+	if err != nil {
+		return nil, err
+	}
+	defer stream.Close()
+
+	req := &share_p2p_v1.GetSharesByNamespaceRequest{
+		RootHash:    root.Hash(),
+		NamespaceId: nID,
+	}
+
+	// if context doesn't have deadline use default one
+	deadline, ok := ctx.Deadline()
+	if !ok {
+		deadline = time.Now().Add(c.defaultTimeout)
+	}
+
+	err = stream.SetDeadline(deadline)
+	if err != nil {
+		log.Debugf("set write deadline: %s", err)
+	}
+
+	_, err = serde.Write(stream, req)
+	if err != nil {
+		return nil, fmt.Errorf("client: writing request: %w", err)
+	}
+
+	err = stream.CloseWrite()
+	if err != nil {
+		log.Debugf("client: closing write side of the stream: %s", err)
+	}
+
+	var resp share_p2p_v1.GetSharesByNamespaceResponse
+	_, err = serde.Read(stream, &resp)
+	if err != nil {
+		return nil, fmt.Errorf("client: reading response: %w", err)
+	}
+
+	if err = statusToErr(resp.Status); err != nil {
+		return nil, fmt.Errorf("client: response code is not OK: %w", err)
+	}
+
+	shares, err := responseToNamespacedShares(resp.Rows)
+	if err != nil {
+		return nil, fmt.Errorf("client: converting response to shares: %w", err)
+	}
+
+	err = shares.Verify(root, nID)
+	if err != nil {
+		return nil, fmt.Errorf("client: verifing response: %w", err)
+	}
+
+	return shares, nil
+}
+
+// responseToNamespacedShares converts proto Rows to share.NamespacedShares
+func responseToNamespacedShares(rows []*share_p2p_v1.Row) (share.NamespacedShares, error) {
+	shares := make([]share.NamespacedRow, 0, len(rows))
+	for _, row := range rows {
+		var proof *ipld.Proof
+		if row.Proof != nil {
+			cids := make([]cid.Cid, 0, len(row.Proof.Nodes))
+			for _, node := range row.Proof.Nodes {
+				cid, err := cid.Cast(node)
+				if err != nil {
+					return nil, fmt.Errorf("casting proofs node to cid: %w", err)
+				}
+				cids = append(cids, cid)
+			}
+
+			proof = &ipld.Proof{
+				Nodes: cids,
+				Start: int(row.Proof.Start),
+				End:   int(row.Proof.End),
+			}
+		}
+
+		shares = append(shares, share.NamespacedRow{
+			Shares: row.Shares,
+			Proof:  proof,
+		})
+	}
+	return shares, nil
+}
+
+func statusToErr(code share_p2p_v1.StatusCode) error {
+	switch code {
+	case share_p2p_v1.StatusCode_OK:
+		return nil
+	case share_p2p_v1.StatusCode_INVALID,
+		share_p2p_v1.StatusCode_NOT_FOUND,
+		share_p2p_v1.StatusCode_INTERNAL,
+		share_p2p_v1.StatusCode_REFUSED:
+		fallthrough
+	default:
+		code = share_p2p_v1.StatusCode_INVALID
```

Comment:
Can you leave only the default case?

Reply:
Fixed. `fallthrough` was a mistake, removed it. I want to have an actual error string being returned, but not "INVALID" all the time.